src/vmm/src/vstate/vcpu/mod.rs

```diff
@@ -624,10 +647,136 @@ mod tests {
+        assert_eq!(
+            format!("{:?}", res.unwrap_err()),
+            format!(
+                "{:?}",
+                EmulationError::FaultyKvmExit("FailEntry".to_string())
+            )
+        );
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Ok(VcpuExit::InternalError));
+        let res = vcpu.run_emulation();
+        assert!(res.is_err());
+        assert_eq!(
+            format!("{:?}", res.unwrap_err()),
+            format!(
+                "{:?}",
+                EmulationError::FaultyKvmExit("InternalError".to_string())
+            )
+        );
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Ok(VcpuExit::SystemEvent(2, 0)));
+        let res = vcpu.run_emulation();
+        assert!(res.is_ok());
+        assert_eq!(res.unwrap(), VcpuEmulation::Stopped);
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Ok(VcpuExit::SystemEvent(1, 0)));
+        let res = vcpu.run_emulation();
+        assert!(res.is_ok());
+        assert_eq!(res.unwrap(), VcpuEmulation::Stopped);
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Ok(VcpuExit::SystemEvent(3, 0)));
+        let res = vcpu.run_emulation();
+        assert!(res.is_err());
+        assert_eq!(
+            format!("{:?}", res.unwrap_err()),
+            format!(
+                "{:?}",
+                EmulationError::FaultyKvmExit("SystemEvent(3, 0)".to_string())
+            )
+        );
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Err(errno::Error::new(libc::EAGAIN)));
+        let res = vcpu.run_emulation();
+        assert!(res.is_ok());
+        assert_eq!(res.unwrap(), VcpuEmulation::Handled);
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Err(errno::Error::new(libc::ENOSYS)));
+        let res = vcpu.run_emulation();
+        assert!(res.is_err());
+        assert_eq!(
+            format!("{:?}", res.unwrap_err()),
+            format!(
+                "{:?}",
+                EmulationError::FaultyKvmExit("Function not implemented (os error 38)".to_string())
+            )
+        );
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Err(errno::Error::new(libc::EINTR)));
+        let res = vcpu.run_emulation();
+        assert!(res.is_ok());
+        assert_eq!(res.unwrap(), VcpuEmulation::Interrupted);
+
+        *(vcpu.vcpu_exit_reason.lock().unwrap()) = Some(Err(errno::Error::new(libc::EINVAL)));
+        let res = vcpu.run_emulation();
+        assert!(res.is_err());
+        assert_eq!(
+            format!("{:?}", res.unwrap_err()),
+            format!(
+                "{:?}",
+                EmulationError::FaultyKvmExit("Invalid argument (os error 22)".to_string())
+            )
+        );
+
+        let mut bus = devices::Bus::new();
+        let dummy = Arc::new(Mutex::new(DummyDevice));
+        bus.insert(dummy, 0x10, 0x10).unwrap();
+        vcpu.set_mmio_bus(bus);
+        let addr = 0x10;
+        static mut DATA: [u8; 4] = [0, 0, 0, 0];
+
+        unsafe {
```

Comment:
I don't think you need the `unsafe` blocks..

Reply:
Sice the `DATA` is static the compiler complains if we do not use the unsafe block
`Use of mutable static is unsafe and requires unsafe function or block`

Reply:
Oh, `MmioRead` takes reference of `'static` lifetime 😞 . Ugly, but c'est la vie.